core/src/miner/stratum.rs

```diff
@@ -0,0 +1,173 @@
+
+        // TODO: should provide initial values for mining
+        None
+    }
+
+    fn submit(&self, payload: (H256, Vec<Bytes>)) -> Result<(), StratumServiceError> {
+        let (pow_hash, seal) = payload;
+
+        ctrace!(STRATUM, "submit_work: Decoded: pow_hash={}, seal={:?}", pow_hash, seal);
+
+        if !self.miner.can_produce_work_package() {
+            cwarn!(STRATUM, "Cannot get work package - engine seals internally.");
+            return Err(StratumServiceError::NoWork)
+        }
+
+        match self.miner.submit_seal(&*self.client, pow_hash, seal) {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                cwarn!(STRATUM, "submit_seal error: {:?}", e);
+                Err(StratumServiceError::Dispatch(e.to_string()))
+            }
+        }
+    }
+}
+
+impl StratumJobDispatcher {
+    /// New stratum job dispatcher given the miner and client
+    fn new(miner: Arc<Miner>, client: Arc<Client>) -> StratumJobDispatcher {
+        StratumJobDispatcher {
+            client,
+            miner,
+        }
+    }
+
+    /// Serializes payload for stratum service
+    fn payload(&self, pow_hash: H256, target: U256) -> String {
+        format!(r#"["0x{:x}","0x{:x}"]"#, pow_hash, target)
+    }
+}
+/// Wrapper for dedicated stratum service
+pub struct Stratum {
+    dispatcher: Arc<StratumJobDispatcher>,
+    service: Arc<StratumService>,
+}
+
+#[derive(Debug)]
+/// Stratum error
+pub enum Error {
+    /// IPC sockets error
+    Service(StratumServiceError),
+    /// Invalid network address
+    Address(AddrParseError),
+}
+
+impl From<StratumServiceError> for Error {
+    fn from(service_err: StratumServiceError) -> Error {
+        Error::Service(service_err)
+    }
+}
+
+impl From<AddrParseError> for Error {
+    fn from(err: AddrParseError) -> Error {
+        Error::Address(err)
+    }
+}
+
+impl NotifyWork for Stratum {
+    fn notify(&self, pow_hash: H256, target: U256) {
+        ctrace!(STRATUM, "Notify work");
+
+        self.service
+            .push_work_all(self.dispatcher.payload(pow_hash, target))
+            .unwrap_or_else(|e| cwarn!(STRATUM, "Error while pushing work: {:?}", e));
+    }
+}
+
+impl Stratum {
+    /// New stratum job dispatcher, given the miner, client and dedicated stratum service
+    pub fn start(config: &Config, miner: Arc<Miner>, client: Arc<Client>) -> Result<Stratum, Error> {
+        use std::net::IpAddr;
+
+        let dispatcher = Arc::new(StratumJobDispatcher::new(miner, client));
+        let stratum_svc = StratumService::start(
+            &SocketAddr::new(config.listen_addr.parse::<IpAddr>()?, config.port),
+            dispatcher.clone(),
+            config.secret.clone(),
+        )?;
+
+        Ok(Stratum {
+            dispatcher,
+            service: stratum_svc,
+        })
+    }
+
+    /// Start STRATUM job dispatcher and register it in the miner
+    pub fn register(cfg: &Config, miner: Arc<Miner>, client: Arc<Client>) -> Result<(), Error> {
```

Comment:
Is there any reason for implementing this here?
This kind of registering operations are usually implemented in `main.rs`.